File: 정수 삼각형.js
// https://www.acmicpc.net/problem/1932

const fs = require('fs')
const input = fs.readFileSync('dev/stdin').toString().trim().split('\n')

const [num, ...tests] = input
const n = +num
const triangle = tests.map((x) => x.split(' ')).map((x) => x.map((x) => +x))

const p = []
for (let i = 0; i < n; i++) {
	const array = []
	for (let j = 0; j <= i; j++) {
		array.push(null)
	}
	p.push(array)
}

p[0][0] = triangle[0][0]

for (let i = 1; i < n; i++) {
	for (let j = 0; j <= i; j++) {
		if (j === 0) {
			p[i][j] = p[i - 1][j] + triangle[i][j]
		} else if (j === i) {
			p[i][j] = p[i - 1][j - 1] + triangle[i][j]
		} else {
			p[i][j] = Math.max(p[i - 1][j - 1], p[i - 1][j]) + triangle[i][j]
		}
	}
}

console.log(Math.max(...p[n - 1]))

File: 이것이 코딩 테스트다/정수 삼각형.py
import sys
sys.stdin = open('./dev/stdin', 'r')
input = sys.stdin.readline

n = int(input())
graph = []
for _ in range(n):
    graph.append(list(map(int, input().split())))

dp = []
for i in range(1, n + 1):
    dp.append([0] * i)

dp[0][0] = graph[0][0]

for i in range(1, n):
    for j in range(i + 1):
        previous_j = [j - 1, j]
        for pj in previous_j:
            if 0 <= pj < i:
                dp[i][j] = max(dp[i][j], dp[i - 1][pj] + graph[i][j])
print(max(dp[-1]))